{"cells":[{"cell_type":"code","execution_count":72,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["[2 2 1 0 1 2 0 1 1 0] <=> GGCACGACCA\n","[1 1 1 1 0 1 2 0 3 3] <=> CCCCACGATT\n","[2 0 2 3 3 1 3 1 3 2] <=> GAGTTCTCTG\n","[0 1 1 1 0 3 1 3 3 3] <=> ACCCATCTTT\n","[2 2 1 0 3 1 1 1 0 1] <=> GGCATCCCAC\n","[1 2 1 2 1 0 2 3 1 2] <=> CGCGCAGTCG\n","[1 3 2 3 0 0 0 0 3 3] <=> CTGTAAAATT\n","[0 3 2 2 0 2 3 1 1 3] <=> ATGGAGTCCT\n","[2 3 1 1 2 3 1 2 0 3] <=> GTCCGTCGAT\n","[0 3 0 0 1 3 0 3 3 1] <=> ATAACTATTC\n"]}],"source":["import numpy as np\n","from pprint import pprint\n","\n","nucleotides_id_mapping = {'A': 0, 'C': 1, 'G': 2, 'T': 3}\n","id_nucleotides_mapping = {value: key for key,\n","                          value in nucleotides_id_mapping.items()}\n","\n","\n","def nucleotidesToId(nucleotidesString: str):\n","    nucleotidesCharArray = np.array(list(nucleotidesString))\n","    nucleotidesIntArray = np.vectorize(\n","        nucleotides_id_mapping.get)(nucleotidesCharArray)\n","    nucleotidesIdArray = nucleotidesIntArray.astype('uint8')\n","    return nucleotidesIdArray\n","\n","\n","def idToNucleotides(idArray):\n","    nucleotidesStringArray = np.vectorize(id_nucleotides_mapping.get)(idArray)\n","    nucleotidesString = \"\".join(nucleotidesStringArray)\n","    return nucleotidesString\n","\n","\n","# Testing\n","for _ in range(10):\n","    randomArray = np.random.choice(list(range(4)), 10)\n","    nucleotidesString = idToNucleotides(randomArray)\n","    # print(f'{randomArray} <=> {nucleotidesString}')\n","    assert((randomArray == nucleotidesToId(nucleotidesString)).all())\n"]},{"cell_type":"code","execution_count":71,"metadata":{},"outputs":[{"data":{"text/plain":["'TGCCAGT'"]},"execution_count":71,"metadata":{},"output_type":"execute_result"}],"source":["def complementaryStrand(nucleotides: str):\n","    nucleotidesIdArray = nucleotidesToId(nucleotides)\n","    complementaryStrandIdArray = 3 - nucleotidesIdArray\n","    return idToNucleotides(complementaryStrandIdArray)\n","\n","complementaryStrand(\"ACGGTCA\")"]},{"cell_type":"code","execution_count":59,"metadata":{},"outputs":[{"data":{"text/plain":["[0, 1, 2, 3, 4]"]},"execution_count":59,"metadata":{},"output_type":"execute_result"}],"source":["list(range(5))"]},{"cell_type":"markdown","metadata":{"id":"kKjeX7krJ0sR"},"source":["# Counting DNA Nucleotides"]},{"cell_type":"code","execution_count":13,"metadata":{"executionInfo":{"elapsed":899,"status":"ok","timestamp":1614626161852,"user":{"displayName":"Agustín Díaz","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gi_EaZNPSfJF9LhuCLRbq1nxucJEfElhyQqAe2M=s64","userId":"07603046028341948051"},"user_tz":180},"id":"JfJh4LD8Ka8m"},"outputs":[],"source":["def countDNANucleotides(DNAString):\n","  counter = {\"A\": 0, \"C\": 0, \"G\": 0, \"T\": 0}\n","  for nucleicAcid in DNAString: \n","    counter[nucleicAcid] += 1\n","        \n","  return counter[\"A\"], counter[\"C\"], counter[\"G\"], counter[\"T\"]\n"]},{"cell_type":"code","execution_count":14,"metadata":{"executionInfo":{"elapsed":836,"status":"ok","timestamp":1614626392222,"user":{"displayName":"Agustín Díaz","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gi_EaZNPSfJF9LhuCLRbq1nxucJEfElhyQqAe2M=s64","userId":"07603046028341948051"},"user_tz":180},"id":"iE3hGl6xMmNt"},"outputs":[],"source":["sampleDNAString = \"GTAGTACCACTCTAATTTTCTGTAGCGCATTGGTGGATAATAGGAGCTCCCTGATGCCATGGAAGCATTCCAAAGCCCCTCGTCCCATAAGTGGACAAGTCAGAATGTCCGGAGCATACTTCGCTGTTCAATGGCCGACAAGTCCGCCGAAGTCTGTTTAAGTCTCGTCACCAACCCCTTAGGACAAGCCCCGGTTCAATACTACCGGTCAGATCCAGAATCTCGACCTCCCGACCTCGCGTACACCCTGTATCTAGGTATCGTACCGCCGCCGGGCACGTTCTTTCCGGATGGGGACGAATGTATGAACATAAACGTCAGATCTAGGAAAATTGTACGATTAGGGCGACTTAATGCGAATCAATAGTTAGGCCCGCACTGCGTTTAAACTTCATTATCTATGTGACGTCATTATCGTTCCTTACGGATCGTAAACGTTTACACGCGGGCCGTGGGTCATAATGTAGCACGTCAAGCGCGAAGTTCTGTTATAAGGGAAACTACCCGTATGACTTTTTACGATGAAGCCTAAAAACCTTTTGGATCGGATGTAGGAGGTAGTTGGCATCCCCGAACAGACCTGGTTGCCCAAGACTACGCGACAATGGTAATCTGAGGTGACTGCTTATGGCATGGGACTTCGGTTTAGGTTTGACATTCAAACTTTACTAATTACTTAACCCGACCGGACTAACACAATCCATTAACGGCCCACCATGATTCCGGGGTTTTCACCTTACTGCCAAGAGATCCAATAGGTGTACGGTTATCGAATAGTTTACAAATCCGTTGTGCGGCTGGTCCGCGTACTGGCTGTTCCTGCTTGATTCTAAGCTGCCATGGAACTTTCGCCGCCTATGTTTCGC\"\n","assert((215, 219, 201, 231) == countDNANucleotides(sampleDNAString))"]},{"cell_type":"markdown","metadata":{"id":"MDtJsnI7O6xa"},"source":["# Transcribing DNA into RNA"]},{"cell_type":"code","execution_count":15,"metadata":{"id":"8RpKPZgPP36A"},"outputs":[],"source":["def DNAtoRNA(DNAString):\n","    RNAString = \"\"\n","    for nucleicAcid in DNAString:\n","        if (nucleicAcid == \"T\"):\n","            RNAString += \"U\"\n","        else:\n","            RNAString += nucleicAcid\n","    return RNAString\n"]},{"cell_type":"code","execution_count":16,"metadata":{"executionInfo":{"elapsed":1274,"status":"ok","timestamp":1614626653078,"user":{"displayName":"Agustín Díaz","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gi_EaZNPSfJF9LhuCLRbq1nxucJEfElhyQqAe2M=s64","userId":"07603046028341948051"},"user_tz":180},"id":"3BGjubtSPks8"},"outputs":[],"source":["sampleDNAString = \"GATGGAACTTGACTACGTAAATT\"\n","assert(\"GAUGGAACUUGACUACGUAAAUU\" == DNAtoRNA(sampleDNAString))\n"]},{"cell_type":"markdown","metadata":{},"source":["# Complementing a Strand of DNA "]},{"cell_type":"code","execution_count":17,"metadata":{},"outputs":[],"source":["def complementDNAStrand(DNAString):\n","    complementaryBases = {\"A\": \"T\", \"T\": \"A\", \"C\": \"G\", \"G\": \"C\"}\n","    complementaryDNAStrand = \"\"\n","    for nucleicAcid in DNAString:\n","        complementaryDNAStrand += complementaryBases[nucleicAcid]\n","\n","    return complementaryDNAStrand[::-1]\n"]},{"cell_type":"code","execution_count":18,"metadata":{},"outputs":[],"source":["assert(\"ACCGGGTTTT\" == complementDNAStrand(\"AAAACCCGGT\"))"]},{"cell_type":"markdown","metadata":{},"source":["# Computing GC Content"]},{"cell_type":"code","execution_count":19,"metadata":{},"outputs":[],"source":["def computeGCContentPercentage(NAString):\n","    GCcounter = 0\n","    for nucleicAcid in NAString:\n","        if nucleicAcid == \"C\":\n","            GCcounter += 1\n","        elif nucleicAcid == \"G\":\n","            GCcounter += 1\n","    return GCcounter / len(NAString) * 100"]},{"cell_type":"code","execution_count":20,"metadata":{},"outputs":[],"source":["assert(computeGCContentPercentage(\"CCACCCTCGTGGTATGGCTAGGCATTCAGGAACCGGAGAACGCTTCAGACCAGCCCGGACTGGGAACCTGCGGGCAGTAGGTGGAAT\") == 60.91954022988506)"]},{"cell_type":"code","execution_count":21,"metadata":{},"outputs":[],"source":["# Uncomment this line if you don't have biopython installed\n","# pip install biopython\n","\n","def highestGCContent(file):\n","    from Bio import SeqIO\n","    maxGCContent = 0\n","    for record in SeqIO.parse(file, \"fasta\"):\n","        recordGC = computeGCContentPercentage(record.seq)\n","        if recordGC > maxGCContent:\n","            name = record.name\n","            maxGCContent = recordGC\n","    \n","    return name, maxGCContent\n"]},{"cell_type":"code","execution_count":22,"metadata":{},"outputs":[{"ename":"ModuleNotFoundError","evalue":"No module named 'Bio'","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)","\u001b[0;32m<ipython-input-22-ff090f7f1c6b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32massert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhighestGCContent\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"f.fasta\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m'Rosalind_0808'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m60.91954022988506\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;32m<ipython-input-21-1671b70c24d4>\u001b[0m in \u001b[0;36mhighestGCContent\u001b[0;34m(file)\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mhighestGCContent\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m     \u001b[0;32mfrom\u001b[0m \u001b[0mBio\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mSeqIO\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m     \u001b[0mmaxGCContent\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mrecord\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mSeqIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'Bio'"]}],"source":["assert(highestGCContent(\"f.fasta\") == ('Rosalind_0808', 60.91954022988506))"]},{"cell_type":"markdown","metadata":{},"source":["# Rabbits and Recurrence Relations"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["def rabbits(n, k):\n","    if n == 1:\n","        return 1\n","    if n == 2:\n","        return 1\n","\n","    serie = [1, 1]\n","    for index in range(2, n):\n","        serie.append(serie[index - 2] * k + serie[index - 1])\n","    return serie[-1]\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["assert(rabbits(28, 4) == 66507086889)"]},{"cell_type":"markdown","metadata":{},"source":["# Counting Point Mutations"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["def calculateHammingDistance (DNAString1, DNAString2):\n","    Hammingcounter = 0\n","    for index in range(len(DNAString1)):\n","        if DNAString1[index] != DNAString2[index]:\n","            Hammingcounter += 1\n","    return Hammingcounter"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["assert(calculateHammingDistance(\"GAGCCTACTAACGGGAT\", \"CATCGTAATGACGGCCT\") == 7)"]},{"cell_type":"markdown","metadata":{},"source":["# Mendel's First Law"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["def dominantAlleleProbabilityFromHomozygousDominant_Heterozygous_HomozygousRecessive(k, m, n):\n","    return (\n","        k * (k - 1) +\n","        2 * k * m + \n","        2 * k * n + \n","        m * (m - 1) * .75+\n","        2 * m * n * .5\n","    ) / (k + m + n) / (k + m + n - 1)\n","        "]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["assert(dominantAlleleProbabilityFromHomozygousDominant_Heterozygous_HomozygousRecessive(26, 29, 24) == 0.7645245050308341)"]},{"cell_type":"markdown","metadata":{},"source":["# Translating RNA into Protein"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[" table = {\n","    'UUU': 'F', 'CUU': 'L', 'AUU': 'I', 'GUU': 'V',\n","    'UUC': 'F', 'CUC': 'L', 'AUC': 'I', 'GUC': 'V',\n","    'UUA': 'L', 'CUA': 'L', 'AUA': 'I', 'GUA': 'V',\n","    'UUG': 'L', 'CUG': 'L', 'AUG': 'M', 'GUG': 'V',\n","    'UCU': 'S', 'CCU': 'P', 'ACU': 'T', 'GCU': 'A',\n","    'UCC': 'S', 'CCC': 'P', 'ACC': 'T', 'GCC': 'A',\n","    'UCA': 'S', 'CCA': 'P', 'ACA': 'T', 'GCA': 'A',\n","    'UCG': 'S', 'CCG': 'P', 'ACG': 'T', 'GCG': 'A',\n","    'UAU': 'Y', 'CAU': 'H', 'AAU': 'N', 'GAU': 'D',\n","    'UAC': 'Y', 'CAC': 'H', 'AAC': 'N', 'GAC': 'D',\n","    'UAA': 'Stop', 'CAA': 'Q', 'AAA': 'K', 'GAA': 'E',\n","    'UAG': 'Stop', 'CAG': 'Q', 'AAG': 'K', 'GAG': 'E',\n","    'UGU': 'C', 'CGU': 'R', 'AGU': 'S', 'GGU': 'G',\n","    'UGC': 'C', 'CGC': 'R', 'AGC': 'S', 'GGC': 'G',\n","    'UGA': '', 'CGA': 'R', 'AGA': 'R', 'GGA': 'G',\n","    'UGG': 'W', 'CGG': 'R', 'AGG': 'R', 'GGG': 'G'\n","}\n","\n","def mRNAtoAminoAcids(mRNAChain):\n","   from textwrap import wrap\n","   codons = wrap(mRNAChain, 3)\n","   assert(codons[0] == 'AUG')\n","   return ''.join([table[codon] for codon in codons])\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["assert('MAMAPRTEINSTRING', mRNAtoAminoAcids('AUGGCCAUGGCGCCCAGAACUGAGAUCAAUAGUACCCGUAUUAACGGGUGA'))"]},{"cell_type":"markdown","metadata":{},"source":["# Finding a Motif in DNA"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["def motifLocationFinder(DNAString, motif):\n","    motifLength = len(motif)\n","    motifLocations = ''\n","    for location in range(len(DNAString) - motifLength + 1): \n","        if DNAString[location:location + motifLength] == motif:\n","            motifLocations += ' ' + str(location + 1)\n","    return motifLocations[1:]"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["assert(motifLocationFinder('GATATATGCATATA', 'ATAT') == '2 4 10')"]},{"cell_type":"markdown","metadata":{},"source":["# Consensus and Profile"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["def mostLikelyCommonAncestorFinder(DNAStrings):\n","    "]}],"metadata":{"colab":{"authorship_tag":"ABX9TyNQqQh75E6nA0kMbhUTwqtH","collapsed_sections":[],"name":"ROSALIND Bioinformatics Stronghold.ipynb","provenance":[],"toc_visible":true},"interpreter":{"hash":"767d51c1340bd893661ea55ea3124f6de3c7a262a8b4abca0554b478b1e2ff90"},"kernelspec":{"display_name":"Python 3.9.5 64-bit ('usr')","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.9.5"}},"nbformat":4,"nbformat_minor":0}