{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"ROSALIND Bioinformatics Stronghold.ipynb","provenance":[],"collapsed_sections":[],"toc_visible":true,"authorship_tag":"ABX9TyNQqQh75E6nA0kMbhUTwqtH"},"kernelspec":{"name":"python3","display_name":"Python 3.10.0 64-bit","metadata":{"interpreter":{"hash":"e332430289709d6a6dda9614c0d652f3460e308bdd8240fb82642a35020f304d"}}}},"cells":[{"cell_type":"markdown","metadata":{"id":"kKjeX7krJ0sR"},"source":["# Counting DNA Nucleotides"]},{"cell_type":"code","metadata":{"id":"JfJh4LD8Ka8m","executionInfo":{"status":"ok","timestamp":1614626161852,"user_tz":180,"elapsed":899,"user":{"displayName":"Agustín Díaz","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gi_EaZNPSfJF9LhuCLRbq1nxucJEfElhyQqAe2M=s64","userId":"07603046028341948051"}}},"source":["def countDNANucleotides(DNAString):\n","  counter = {\"A\": 0, \"C\": 0, \"G\": 0, \"T\": 0}\n","  for nucleicAcid in DNAString: \n","    counter[nucleicAcid] += 1\n","        \n","  return counter[\"A\"], counter[\"C\"], counter[\"G\"], counter[\"T\"]\n"],"execution_count":2,"outputs":[]},{"cell_type":"code","metadata":{"id":"iE3hGl6xMmNt","executionInfo":{"status":"ok","timestamp":1614626392222,"user_tz":180,"elapsed":836,"user":{"displayName":"Agustín Díaz","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gi_EaZNPSfJF9LhuCLRbq1nxucJEfElhyQqAe2M=s64","userId":"07603046028341948051"}}},"source":["sampleDNAString = \"GTAGTACCACTCTAATTTTCTGTAGCGCATTGGTGGATAATAGGAGCTCCCTGATGCCATGGAAGCATTCCAAAGCCCCTCGTCCCATAAGTGGACAAGTCAGAATGTCCGGAGCATACTTCGCTGTTCAATGGCCGACAAGTCCGCCGAAGTCTGTTTAAGTCTCGTCACCAACCCCTTAGGACAAGCCCCGGTTCAATACTACCGGTCAGATCCAGAATCTCGACCTCCCGACCTCGCGTACACCCTGTATCTAGGTATCGTACCGCCGCCGGGCACGTTCTTTCCGGATGGGGACGAATGTATGAACATAAACGTCAGATCTAGGAAAATTGTACGATTAGGGCGACTTAATGCGAATCAATAGTTAGGCCCGCACTGCGTTTAAACTTCATTATCTATGTGACGTCATTATCGTTCCTTACGGATCGTAAACGTTTACACGCGGGCCGTGGGTCATAATGTAGCACGTCAAGCGCGAAGTTCTGTTATAAGGGAAACTACCCGTATGACTTTTTACGATGAAGCCTAAAAACCTTTTGGATCGGATGTAGGAGGTAGTTGGCATCCCCGAACAGACCTGGTTGCCCAAGACTACGCGACAATGGTAATCTGAGGTGACTGCTTATGGCATGGGACTTCGGTTTAGGTTTGACATTCAAACTTTACTAATTACTTAACCCGACCGGACTAACACAATCCATTAACGGCCCACCATGATTCCGGGGTTTTCACCTTACTGCCAAGAGATCCAATAGGTGTACGGTTATCGAATAGTTTACAAATCCGTTGTGCGGCTGGTCCGCGTACTGGCTGTTCCTGCTTGATTCTAAGCTGCCATGGAACTTTCGCCGCCTATGTTTCGC\"\n","assert((215, 219, 201, 231) == countDNANucleotides(sampleDNAString))"],"execution_count":3,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"MDtJsnI7O6xa"},"source":["# Transcribing DNA into RNA"]},{"cell_type":"code","metadata":{"id":"8RpKPZgPP36A"},"source":["def DNAtoRNA(DNAString):\n","    RNAString = \"\"\n","    for nucleicAcid in DNAString:\n","        if (nucleicAcid == \"T\"):\n","            RNAString += \"U\"\n","        else: \n","            RNAString += nucleicAcid\n","    return RNAString"],"execution_count":6,"outputs":[]},{"cell_type":"code","metadata":{"id":"3BGjubtSPks8","executionInfo":{"status":"ok","timestamp":1614626653078,"user_tz":180,"elapsed":1274,"user":{"displayName":"Agustín Díaz","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gi_EaZNPSfJF9LhuCLRbq1nxucJEfElhyQqAe2M=s64","userId":"07603046028341948051"}}},"source":["sampleDNAString = \"GATGGAACTTGACTACGTAAATT\"\n","assert(\"GAUGGAACUUGACUACGUAAAUU\" == DNAtoRNA(sampleDNAString))"],"execution_count":7,"outputs":[]},{"source":["# Complementing a Strand of DNA "],"cell_type":"markdown","metadata":{}},{"cell_type":"code","execution_count":8,"metadata":{},"outputs":[],"source":["def complementDNAStrand (DNAString):\n","    complementaryBases= {\"A\": \"T\", \"T\": \"A\", \"C\": \"G\", \"G\": \"C\"}\n","    complementaryDNAStrand = \"\" \n","    for nucleicAcid in DNAString:\n","        complementaryDNAStrand += complementaryBases[nucleicAcid]\n","\n","    return complementaryDNAStrand[::-1]"]},{"cell_type":"code","execution_count":9,"metadata":{},"outputs":[],"source":["assert(\"ACCGGGTTTT\" == complementDNAStrand(\"AAAACCCGGT\"))"]},{"source":["# Computing GC Content"],"cell_type":"markdown","metadata":{}},{"cell_type":"code","execution_count":17,"metadata":{},"outputs":[],"source":["def computeGCContentPercentage(NAString):\n","    GCcounter = 0\n","    for nucleicAcid in NAString:\n","        if nucleicAcid == \"C\":\n","            GCcounter += 1\n","        elif nucleicAcid == \"G\":\n","            GCcounter += 1\n","    return GCcounter / len(NAString) * 100"]},{"cell_type":"code","execution_count":22,"metadata":{},"outputs":[],"source":["assert(computeGCContentPercentage(\"CCACCCTCGTGGTATGGCTAGGCATTCAGGAACCGGAGAACGCTTCAGACCAGCCCGGACTGGGAACCTGCGGGCAGTAGGTGGAAT\") == 60.91954022988506)"]},{"cell_type":"code","execution_count":23,"metadata":{},"outputs":[],"source":["# Uncomment this line if you don't have biopython installed\n","# pip install biopython\n","\n","def highestGCContent(file):\n","    from Bio import SeqIO\n","    maxGCContent = 0\n","    for record in SeqIO.parse(file, \"fasta\"):\n","        recordGC = computeGCContentPercentage(record.seq)\n","        if recordGC > maxGCContent:\n","            name = record.name\n","            maxGCContent = recordGC\n","    \n","    return name, maxGCContent\n"]},{"cell_type":"code","execution_count":25,"metadata":{},"outputs":[],"source":["assert(highestGCContent(\"f.fasta\") == ('Rosalind_0808', 60.91954022988506))"]},{"source":["# Rabbits and Recurrence Relations"],"cell_type":"markdown","metadata":{}},{"cell_type":"code","execution_count":51,"metadata":{},"outputs":[],"source":["def rabbits(n, k):\n","    if n == 1:\n","        return 1\n","    if n == 2:\n","        return 1\n","\n","    serie = [1, 1]\n","    for index in range(2, n):\n","        serie.append(serie[index - 2] * k + serie[index - 1])\n","    return serie[-1]\n"]},{"cell_type":"code","execution_count":52,"metadata":{},"outputs":[],"source":["assert(rabbits(28, 4) == 66507086889)"]},{"source":["# Counting Point Mutations"],"cell_type":"markdown","metadata":{}},{"cell_type":"code","execution_count":2,"metadata":{},"outputs":[],"source":["def calculateHammingDistance (DNAString1, DNAString2):\n","    Hammingcounter = 0\n","    for index in range(len(DNAString1)):\n","        if DNAString1[index] != DNAString2[index]:\n","            Hammingcounter += 1\n","    return Hammingcounter"]},{"cell_type":"code","execution_count":6,"metadata":{},"outputs":[],"source":["assert(calculateHammingDistance(\"GAGCCTACTAACGGGAT\", \"CATCGTAATGACGGCCT\") == 7)"]},{"source":["# Mendel's First Law"],"cell_type":"markdown","metadata":{}},{"cell_type":"code","execution_count":20,"metadata":{},"outputs":[],"source":["def dominantAlleleProbabilityFromHomozygousDominant_Heterozygous_HomozygousRecessive(k, m, n):\n","    return (\n","        k * (k - 1) +\n","        2 * k * m + \n","        2 * k * n + \n","        m * (m - 1) * .75+\n","        2 * m * n * .5\n","    ) / (k + m + n) / (k + m + n - 1)\n","        "]},{"cell_type":"code","execution_count":23,"metadata":{},"outputs":[],"source":["assert(dominantAlleleProbabilityFromHomozygousDominant_Heterozygous_HomozygousRecessive(26, 29, 24) == 0.7645245050308341)"]},{"source":["# Translating RNA into Protein"],"cell_type":"markdown","metadata":{}},{"cell_type":"code","execution_count":9,"metadata":{},"outputs":[],"source":[" table = {\n","    'UUU': 'F', 'CUU': 'L', 'AUU': 'I', 'GUU': 'V',\n","    'UUC': 'F', 'CUC': 'L', 'AUC': 'I', 'GUC': 'V',\n","    'UUA': 'L', 'CUA': 'L', 'AUA': 'I', 'GUA': 'V',\n","    'UUG': 'L', 'CUG': 'L', 'AUG': 'M', 'GUG': 'V',\n","    'UCU': 'S', 'CCU': 'P', 'ACU': 'T', 'GCU': 'A',\n","    'UCC': 'S', 'CCC': 'P', 'ACC': 'T', 'GCC': 'A',\n","    'UCA': 'S', 'CCA': 'P', 'ACA': 'T', 'GCA': 'A',\n","    'UCG': 'S', 'CCG': 'P', 'ACG': 'T', 'GCG': 'A',\n","    'UAU': 'Y', 'CAU': 'H', 'AAU': 'N', 'GAU': 'D',\n","    'UAC': 'Y', 'CAC': 'H', 'AAC': 'N', 'GAC': 'D',\n","    'UAA': 'Stop', 'CAA': 'Q', 'AAA': 'K', 'GAA': 'E',\n","    'UAG': 'Stop', 'CAG': 'Q', 'AAG': 'K', 'GAG': 'E',\n","    'UGU': 'C', 'CGU': 'R', 'AGU': 'S', 'GGU': 'G',\n","    'UGC': 'C', 'CGC': 'R', 'AGC': 'S', 'GGC': 'G',\n","    'UGA': '', 'CGA': 'R', 'AGA': 'R', 'GGA': 'G',\n","    'UGG': 'W', 'CGG': 'R', 'AGG': 'R', 'GGG': 'G'\n","}\n","\n","def mRNAtoAminoAcids(mRNAChain):\n","   from textwrap import wrap\n","   codons = wrap(mRNAChain, 3)\n","   assert(codons[0] == 'AUG')\n","   return ''.join([table[codon] for codon in codons])\n"]},{"cell_type":"code","execution_count":11,"metadata":{},"outputs":[],"source":["assert('MAMAPRTEINSTRING', mRNAtoAminoAcids('AUGGCCAUGGCGCCCAGAACUGAGAUCAAUAGUACCCGUAUUAACGGGUGA'))"]},{"source":["# Finding a Motif in DNA"],"cell_type":"markdown","metadata":{}},{"cell_type":"code","execution_count":35,"metadata":{},"outputs":[],"source":["def motifLocationFinder(DNAString, motif):\n","    motifLength = len(motif)\n","    motifLocations = ''\n","    for location in range(len(DNAString) - motifLength + 1): \n","        if DNAString[location:location + motifLength] == motif:\n","            motifLocations += ' ' + str(location + 1)\n","    return motifLocations[1:]"]},{"cell_type":"code","execution_count":37,"metadata":{},"outputs":[],"source":["assert(motifLocationFinder('GATATATGCATATA', 'ATAT') == '2 4 10')"]},{"source":["# Consensus and Profile"],"cell_type":"markdown","metadata":{}},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["def mostLikelyCommonAncestorFinder(DNAStrings):\n","    "]}]}